modules/light-clients/07-tendermint/types/proposal_handle.go

```diff
@@ -80,37 +59,33 @@ func (cs ClientState) CheckSubstituteAndUpdateState(
 
 	// copy consensus states and processed time from substitute to subject
 	// starting from initial height and ending on the latest height (inclusive)
-	for i := initialHeight.GetRevisionHeight(); i <= substituteClientState.GetLatestHeight().GetRevisionHeight(); i++ {
-		height := clienttypes.NewHeight(substituteClientState.GetLatestHeight().GetRevisionNumber(), i)
-
-		consensusState, err := GetConsensusState(substituteClientStore, cdc, height)
-		if err != nil {
-			// not all consensus states will be filled in
-			continue
-		}
-		SetConsensusState(subjectClientStore, cdc, consensusState, height)
+	height := substituteClientState.GetLatestHeight()
 
-		// set metadata for this consensus state
-		setConsensusMetadata(ctx, subjectClientStore, height)
+	consensusState, err := GetConsensusState(substituteClientStore, cdc, height)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "unable to retrieve latest consensus state for substitute client")
 	}
 
-	cs.LatestHeight = substituteClientState.LatestHeight
+	SetConsensusState(subjectClientStore, cdc, consensusState, height)
 
-	// validate the updated client and ensure it isn't expired
-	if err := cs.Validate(); err != nil {
-		return nil, sdkerrors.Wrap(err, "unexpected error: updated subject client state is invalid")
+	// set metadata stored for the substitute consensus state
+	processedHeight, found := GetProcessedHeight(substituteClientStore, height)
+	if !found {
+		return nil, sdkerrors.Wrap(clienttypes.ErrUpdateClientFailed, "unable to retrieve processed height for substitute client latest height")
 	}
 
-	latestConsensusState, err := GetConsensusState(subjectClientStore, cdc, cs.GetLatestHeight())
-	if err != nil {
-		return nil, sdkerrors.Wrapf(
-			err, "unexpected error: could not get consensus state for updated subject client from clientstore at height: %d", cs.GetLatestHeight(),
-		)
+	processedTime, found := GetProcessedTime(substituteClientStore, height)
+	if !found {
+		return nil, sdkerrors.Wrap(clienttypes.ErrUpdateClientFailed, "unable to retrieve processed time for substitute client latest height")
 	}
 
-	if cs.IsExpired(latestConsensusState.Timestamp, ctx.BlockTime()) {
-		return nil, sdkerrors.Wrap(clienttypes.ErrInvalidClient, "updated subject client is expired")
-	}
+	setConsensusMetadataWithValues(subjectClientStore, height, processedHeight, processedTime)
```

Comment:
I think you can still just use the `context` here, it just makes things a bit more inconvenient.

Also you need to set the iteration key.

Reply:
The iteration key is being set in [`setConsensusMetadataWithValues`](https://github.com/cosmos/ibc-go/pull/181/files#diff-41404a951cddf32ec70eded59b7e20092d750c3aba4cfa3778b3d31e0e88119dR309)

What's the reasoning for using `setConsensusMetadata`? Less code? It seems odd to me that the substitute and subject wouldn't end up with identical information for the copied consensus state. It also seems odd that this consensus state may not be possible to be verified against (if the delay period > timeToExpiration)

It seems ok to use `setConsensusMetadata`, but I'd like to have some reason for doing and document that since it isn't intuitive and requires extra reasoning

Reply:
No reason apart from less code, approving 👍